libsnark/polynomial_commitments/kzg10_batched.tcc

```diff
@@ -0,0 +1,285 @@
+            other.end(),
+            poly.begin() + p_orig_size);
+    }
+
+    // Add coefficients of `other` into `polynomial`.
+    for (size_t i = 0; i < other.size(); ++i) {
+        poly[i] += other[i];
+    }
+}
+
+template<typename FieldT>
+static polynomial<FieldT> polynomial_accumulate_with_power_factors(
+    const std::vector<polynomial<FieldT>> &polys, const FieldT &factor)
+{
+    const size_t t = polys.size();
+    if (t == 1) {
+        return polys[0];
+    }
+
+    // Start with
+    //   f_accum = polys[t-2] + factor * polys[t-1].
+    //
+    // For each remaining polynomial f:
+    //   f_accum <- factor * f_accum + f
+
+    size_t i = t - 1;
+    polynomial<FieldT> f_accum = polynomial_scalar_mul(polys[i], factor);
+    polynomial_add_inplace(f_accum, polys[--i]);
+    while (i > 0) {
+        polynomial_scalar_mul_inplace(f_accum, factor);
+        polynomial_add_inplace(f_accum, polys[--i]);
+    }
+
+    return f_accum;
+}
+
+} // namespace internal
+
+template<typename ppT>
+kzg10_batched_2_point<ppT>::evaluations::evaluations(
+    std::vector<Field> &&s_1s, std::vector<Field> &&s_2s)
+    : s_1s(s_1s), s_2s(s_2s)
+{
+}
+
+template<typename ppT>
+kzg10_batched_2_point<ppT>::evaluation_witness::evaluation_witness(
+    const libff::G1<ppT> &W_1, const libff::G1<ppT> &W_2)
+    : W_1(W_1), W_2(W_2)
+{
+}
+
+template<typename ppT>
+typename kzg10_batched_2_point<ppT>::evaluations kzg10_batched_2_point<ppT>::
+    evaluate_polynomials(
+        const std::vector<polynomial<Field>> &fs,
+        const std::vector<polynomial<Field>> &gs,
+        const Field &z_1,
+        const Field &z_2)
+{
+    // Denote the numbers of polynomials as t1 and t2, consistent with [GWC19].
+    const size_t t1 = fs.size();
+    const size_t t2 = gs.size();
+
+    std::vector<Field> s_1s;
+    s_1s.reserve(t1);
+    for (const polynomial<Field> &f_i : fs) {
+        s_1s.push_back(libfqfft::evaluate_polynomial(f_i.size(), f_i, z_1));
+    }
+
+    std::vector<Field> s_2s;
+    s_2s.reserve(t2);
+    for (const polynomial<Field> &g_i : gs) {
+        s_2s.push_back(libfqfft::evaluate_polynomial(g_i.size(), g_i, z_2));
+    }
+
+    return evaluations(std::move(s_1s), std::move(s_2s));
+}
+
+template<typename ppT>
+typename kzg10_batched_2_point<ppT>::evaluation_witness kzg10_batched_2_point<
+    ppT>::
+    create_evaluation_witness(
+        const std::vector<polynomial<Field>> &fs,
+        const std::vector<polynomial<Field>> &gs,
+        const Field &z_1,
+        const Field &z_2,
+        const evaluations &evaluations,
+        const srs &srs,
+        const Field &gamma_1,
+        const Field &gamma_2)
+{
+    assert(fs.size() == evaluations.s_1s.size());
+    assert(gs.size() == evaluations.s_2s.size());
+    libff::UNUSED(evaluations);
+
+    // For convenience of variable naming, let $f_i \in fs$ and $g_i in gs$ be
+    // the two sets of polynomials. These are denoted $f_i$ and $f^\prime_i$ in
```

Comment:
That's a detail but I'd use `$f'_i$` which is the same thing pretty much but less heavy notation-wise.

Reply:
Same for the comments below.

Reply:
changed